File: src/local_assets_engine/recipes/previz.py
"""Previz recipe: blockout scene → camera rigs → sketch renders → shot records.

The product is the shot values, not the pictures. A genre is a shot preset, so
this recipe has no trailer-specific code: it resolves the preset's relative
framing against the placed assets and records what the camera actually did.
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path
from typing import TYPE_CHECKING, Any

from ..jobs import JobNotFound
from ..presets import PresetError, find_shot_preset, find_standin
from .base import Recipe, bool_param, choice_param, int_param

if TYPE_CHECKING:
    from ..jobs import JobStore
    from ..runner import JobContext

MESH_SUFFIXES = {".glb", ".gltf"}
MAX_ASSETS = 8
MAX_SHOTS = 24
DEFAULT_PRESET = "game-trailer"
# 샷 id는 결과 폴더 이름이 되고 에셋 id는 Blender 오브젝트 이름이 된다. 경로 문자를 받지 않는다.
SAFE_ID = re.compile(r"^[A-Za-z0-9_-]{1,32}$")
FRAMING_RANGES = {
    "distance": (0.2, 20.0), "azimuth": (-720.0, 720.0), "height": (-1.0, 10.0),
    "targetHeight": (-1.0, 5.0), "roll": (-90.0, 90.0),
}


def _number(value: Any, name: str, low: float, high: float, default: float) -> float:
    if value in (None, ""):
        return default
    try:
        number = float(value)
    except (TypeError, ValueError) as error:
        raise PresetError(f"{name}는 숫자여야 합니다.") from error
    if not low <= number <= high:
        raise PresetError(f"{name}는 {low}~{high} 사이여야 합니다.")
    return number


def _standin(raw: dict[str, Any], presets: dict[str, Any]) -> dict[str, Any]:
    """A gray stand-in from the catalog, stretched per axis to the size asked for."""
    kind = find_standin(presets, str(raw["standin"]))
    given = raw.get("size") or kind["size"]
    if not isinstance(given, list) or len(given) != 3:
        raise PresetError("대역 size는 [가로, 깊이, 높이] 세 값이어야 합니다.")
    size = [_number(value, "대역 size", 0.05, 500.0, float(base)) for value, base in zip(given, kind["size"])]
    stretch = [new / float(base) for new, base in zip(size, kind["size"])]
    # 부품은 카탈로그 치수 기준 미터다. 늘인 결과를 기록에 남겨 카탈로그가 바뀌어도 다시 그릴 수 있다.
    parts = [{
        "shape": part["shape"], "axis": part.get("axis", "z"),
        "at": [round(float(value) * factor, 4) for value, factor in zip(part["at"], stretch)],
        "size": [round(float(value) * factor, 4) for value, factor in zip(part["size"], stretch)],
    } for part in kind["parts"]]
    return {"standin": kind["id"], "label": f"{kind['label']} 대역", "size": size, "parts": parts}


def _placement(raw: Any, index: int, store: "JobStore", presets: dict[str, Any]) -> dict[str, Any]:
    """One thing in the scene: a finished mesh job, a GLB file on disk, or a stand-in."""
    if not isinstance(raw, dict):
        raise PresetError("배치 항목 형식이 올바르지 않습니다.")
    source = raw.get("source")
    standin: dict[str, Any] = {}
    file: Path | None = None
    reference: dict[str, str] | None = None
    if raw.get("standin"):
        standin = _standin(raw, presets)
        label = standin.pop("label")
    elif source:
        job_id, asset_id = str(source.get("jobId")), str(source.get("assetId"))
        try:
            job = store.load(job_id)
            asset = next(a for a in job["assets"] if a["id"] == asset_id and a["kind"] == "mesh")
            file = store.resolve_file(job_id, asset["file"])
        except (JobNotFound, StopIteration) as error:
            raise PresetError("장면에 놓을 3D 에셋을 찾을 수 없습니다.") from error
        label = str(job.get("params", {}).get("subject") or job.get("title") or asset_id)
        reference = {"jobId": job_id, "assetId": asset_id}
    else:
        given = str(raw.get("path") or "")
        file = Path(given).expanduser()
        if not given or not file.is_absolute() or not file.is_file() or file.suffix.lower() not in MESH_SUFFIXES:
            raise PresetError("GLB·glTF 파일의 절대 경로가 필요합니다.")
        label = file.stem
    position = raw.get("position") or (0.0, 0.0, 0.0)
    if len(position) != 3:
        raise PresetError("position은 [x, y, z] 세 값이어야 합니다.")
    asset_id = str(raw.get("id") or ("hero" if index == 0 else f"asset{index + 1}"))
    if not SAFE_ID.match(asset_id):
        raise PresetError("에셋 id는 영문·숫자·-·_ 32자 이내여야 합니다.")
    return {
        "id": asset_id,
        "label": label,
        "file": str(file) if file else None,
        "source": reference,
        **standin,
        "position": [_number(value, "position", -1000.0, 1000.0, 0.0) for value in position],
        "yaw": _number(raw.get("yaw"), "yaw", -360.0, 360.0, 0.0),
        "scale": _number(raw.get("scale"), "scale", 0.01, 100.0, 1.0),
    }


def _edited_shots(raw: Any, known: set[str], hero: str) -> list[dict[str, Any]]:
    """Shots a person rearranged or re-aimed in the app, checked like preset values."""
    if not isinstance(raw, list) or not 1 <= len(raw) <= MAX_SHOTS:
        raise PresetError(f"컷은 1~{MAX_SHOTS}개여야 합니다.")
    shots: list[dict[str, Any]] = []
    for order, item in enumerate(raw, start=1):
        if not isinstance(item, dict):
            raise PresetError("컷 형식이 올바르지 않습니다.")
        shot_id = str(item.get("id") or "")
        if not SAFE_ID.match(shot_id) or any(shot["id"] == shot_id for shot in shots):
            raise PresetError("컷 id는 영문·숫자·-·_ 32자 이내이고 겹치지 않아야 합니다.")
        given = item.get("framing") if isinstance(item.get("framing"), dict) else {}
        framing: dict[str, Any] = {}
        for key, (low, high) in FRAMING_RANGES.items():
            for name in (key, f"{key}End"):
                if given.get(name) not in (None, ""):
                    framing[name] = _number(given[name], f"{shot_id} {name}", low, high, 0.0)
        if given.get("targetOffset") is not None:
            offset = given["targetOffset"]
            if not isinstance(offset, list) or len(offset) != 2:
                raise PresetError(f"{shot_id} targetOffset은 [x, y] 두 값이어야 합니다.")
            framing["targetOffset"] = [_number(value, f"{shot_id} targetOffset", -5.0, 5.0, 0.0) for value in offset]
        focus = str(item.get("focus") or "scene")
        focus = hero if focus == "hero" else focus
        if focus not in known:
            raise PresetError(f"{shot_id}의 focus를 장면에서 찾을 수 없습니다: {focus}")
        shot = {
            "id": shot_id,
            "label": str(item.get("label") or shot_id)[:40],
            "purpose": str(item.get("purpose") or "")[:80],
            "move": str(item.get("move") or "static")[:24],
            "focus": focus,
            "lens": _number(item.get("lens"), f"{shot_id} lens", 8.0, 300.0, 35.0),
            "seconds": _number(item.get("seconds"), f"{shot_id} seconds", 0.2, 60.0, 2.0),
            "ease": "linear" if item.get("ease") == "linear" else "inout",
            "framing": framing,
            "order": order,
        }
        if item.get("lensEnd") not in (None, ""):
            shot["lensEnd"] = _number(item["lensEnd"], f"{shot_id} lensEnd", 8.0, 300.0, shot["lens"])
        shots.append(shot)
    return shots


def _prepare(params: dict[str, Any], presets: dict[str, Any], store: "JobStore") -> tuple[dict[str, Any], str]:
    previz = presets["previz"]
    defaults = previz["defaults"]
    entries = params.get("assets")
    if not entries and (params.get("source") or params.get("path")):
        entries = [{"source": params.get("source"), "path": params.get("path")}]
    if not entries:
        raise PresetError("장면에 놓을 3D 에셋이나 대역이 필요합니다.")
    if not isinstance(entries, list):
        raise PresetError("assets는 배치 항목의 목록이어야 합니다.")
    if len(entries) > MAX_ASSETS:
        raise PresetError(f"한 장면에 에셋과 대역은 합쳐서 {MAX_ASSETS}개까지 놓을 수 있습니다.")
    assets = [_placement(entry, index, store, presets) for index, entry in enumerate(entries)]
    if len({asset["id"] for asset in assets}) != len(assets):
        raise PresetError("에셋 id가 중복됩니다.")

    preset = find_shot_preset(presets, str(params.get("preset") or DEFAULT_PRESET))
    # 프리셋은 "hero"라는 이름으로 주인공을 가리킨다. 장면의 첫 에셋이 그 자리다.
    known = {asset["id"] for asset in assets} | {"scene"}
    hero = assets[0]["id"]
    edited = bool(params.get("shots"))
    if edited:
        shots = _edited_shots(params["shots"], known, hero)
    else:
        shots = []
        for order, shot in enumerate(preset["shots"], start=1):
            focus = str(shot.get("focus", "scene"))
            if focus == "hero":
                focus = hero
            if focus not in known:
                raise PresetError(f"{shot['id']}의 focus를 장면에서 찾을 수 없습니다: {focus}")
            shots.append({**shot, "focus": focus, "order": order})

    clay = bool_param(params, "clay", defaults["clay"])
    normalized = {
        "preset": preset["id"],
        "edited": edited,
        "assets": assets,
        "shots": shots,
        "hero": hero,
        "renderer": choice_param(params, "renderer", defaults["renderer"], previz["renderers"]),
        "width": int_param(params, "width", defaults["width"], 256, 1920),
        "height": int_param(params, "height", defaults["height"], 144, 1080),
        "fps": int_param(params, "fps", defaults["fps"], 6, 30),
        "samples": int_param(params, "samples", defaults["samples"], 1, 256),
        "aux": choice_param(params, "aux", defaults["aux"], previz["auxModes"]),
        "animatic": bool_param(params, "animatic", defaults["animatic"]),
        "ground": bool_param(params, "ground", defaults["ground"]),
        "clay": clay,
        "look": {**preset.get("look", {}), "clay": clay},
    }
    label = f"{preset['label']} 편집" if edited else preset["label"]
    return normalized, f"프리비즈 · {label} · {assets[0]['label']}"


def _run(ctx: "JobContext") -> None:
    p = ctx.params
    previz_dir = ctx.dir / "previz"
    previz_dir.mkdir(exist_ok=True)
    plan_path, result_path = previz_dir / "plan.json", previz_dir / "shots.json"
    plan = {
        key: p[key] for key in
        ("renderer", "width", "height", "fps", "samples", "aux", "animatic", "ground", "look", "shots")
    }
    plan["assets"] = [{key: asset.get(key) for key in ("id", "file", "standin", "parts", "position", "yaw", "scale")}
                      for asset in p["assets"]]
    plan_path.write_text(json.dumps(plan, ensure_ascii=False, indent=2), "utf-8")

    with ctx.stage("previz", "프리비즈 스케치 렌더 (Blender)") as stage:
        stage.progress(0, f"{p['renderer']} · 샷 {len(p['shots'])}개", force=True)
        stage.run([
            sys.executable, "-m", "local_assets_engine.tools.previz_render",
            "--plan", plan_path, "--out-dir", previz_dir, "--result", result_path,
        ], cwd=previz_dir)

    result = json.loads(result_path.read_text("utf-8"))
    sequence = result.get("sequence")
    placement = {cut["shot"]: cut for cut in (sequence or {}).get("cuts", [])}
    for shot in result["shots"]:
        files = shot["files"]
        key_frame = previz_dir / files["key"]
        clip = previz_dir / files["animatic"] if files.get("animatic") else None
        # 컷마다 이어 붙인 한 편 안의 자리를 남긴다. 화면이 그 시점으로 바로 옮겨 간다.
        cut = placement.get(shot["id"])
        in_sequence = {
            "file": ctx.rel(previz_dir / sequence["file"]), "seconds": sequence["seconds"],
            "at": cut["at"], "start": cut["start"], "end": cut["end"],
        } if sequence and cut else None
        ctx.add_asset(
            kind="shot", role="candidate", file=clip or key_frame, preview=key_frame,
            meta={
                "shot": shot["id"], "label": shot["label"], "purpose": shot["purpose"],
                "move": shot["move"], "focus": shot["focus"], "order": shot["order"],
                "seconds": shot["seconds"], "fps": shot["fps"], "frames": shot["frames"],
                "ease": shot["ease"], "lens": shot["lens"], "lensEnd": shot["lensEnd"],
                "sensorWidth": shot["sensorWidth"], "start": shot["start"], "end": shot["end"],
                "framing": shot["framing"], "depthRange": shot["depthRange"],
                "renderer": result["renderer"], "resolution": result["resolution"],
                "clay": result["clay"], "renderSeconds": shot["renderSeconds"],
                "preset": p["preset"], "sequence": in_sequence,
                "blendFile": ctx.rel(previz_dir / result["blend"]) if result.get("blend") else None,
                # 프레임별 카메라 값은 job.json을 불리므로 파일에 두고 여기서 가리킨다.
                "pathFile": ctx.rel(result_path),
                "files": {name: ([ctx.rel(previz_dir / item) for item in value]
                                 if isinstance(value, list) else
                                 ctx.rel(previz_dir / value) if value else None)
                          for name, value in files.items()},
            },
        )


PREVIZ = Recipe(id="previz", label="프리비즈 샷", prepare=_prepare, run=_run)

File: electron-app/shared/previz.mjs
// 프리비즈 장면·카메라 계산. 엔진(tools/previz_render.py)과 같은 식이어야
// 화면 지도에서 끈 카메라가 렌더 결과와 같은 자리에 선다. 브라우저에서도 돈다.

export const SENSOR_MM = 36;
export const ASPECT = 16 / 9;
const DEFAULTS = { distance: 1.8, azimuth: -35, height: 0.6, targetHeight: 0.5 };
const LIMITS = { distance: [0.2, 20], lens: [8, 300], seconds: [0.2, 60], size: [0.05, 500] };
export const SIZE_LABELS = ["가로", "깊이", "높이"];

const clamp = (value, [low, high]) => Math.min(high, Math.max(low, value));
const clone = (value) => JSON.parse(JSON.stringify(value));

export function smooth(start, end, fraction, ease = "inout") {
  const t = ease === "inout" ? fraction * fraction * (3 - 2 * fraction) : fraction;
  return start + (end - start) * t;
}

function pair(source, key, fallback) {
  const start = Number(source?.[key] ?? fallback);
  const end = source?.[`${key}End`];
  return [start, end === undefined || end === null || end === "" ? start : Number(end)];
}

// 엔진은 에셋의 바닥 중심을 배치 좌표에 두고 회전한 뒤 월드 경계 상자를 잰다.
export function footprint(item) {
  const scale = Number(item.scale) || 1;
  const [dx, dy, dz] = (item.dimensions || [1, 1, 1]).map((value) => Number(value) * scale);
  const yaw = ((Number(item.yaw) || 0) * Math.PI) / 180;
  const ex = Math.abs(dx * Math.cos(yaw)) + Math.abs(dy * Math.sin(yaw));
  const ey = Math.abs(dx * Math.sin(yaw)) + Math.abs(dy * Math.cos(yaw));
  const x = Number(item.x) || 0;
  const y = Number(item.y) || 0;
  return { low: [x - ex / 2, y - ey / 2, 0], high: [x + ex / 2, y + ey / 2, dz] };
}

function subjectOf(low, high) {
  return {
    center: [(low[0] + high[0]) / 2, (low[1] + high[1]) / 2],
    baseZ: low[2],
    width: Math.max(high[0] - low[0], high[1] - low[1], 0.05),
    height: Math.max(high[2] - low[2], 0.05),
    radius: Math.max(Math.hypot(high[0] - low[0], high[1] - low[1], high[2] - low[2]) / 2, 0.05),
  };
}

export function assetId(index) {
  return index === 0 ? "hero" : `asset${index + 1}`;
}

// 배치 순서가 바뀌면 에셋 id(asset2…)가 가리키는 물체도 바뀐다. order[새 자리] = 옛 자리로 받아
// 컷이 같은 물체를 계속 겨냥하게 옮긴다. hero·scene은 역할이라 그대로 두고, 빠진 물체를 보던 컷은 장면 전체를 본다.
export function refocusCuts(cuts, order) {
  const now = new Map(order.map((old, index) => [old, index]));
  return cuts.map((cut) => {
    const match = /^asset(\d+)$/.exec(cut.focus || "");
    if (!match) return cut;
    const index = now.get(Number(match[1]) - 1);
    return { ...cut, focus: index === undefined ? "scene" : assetId(index) };
  });
}

// 새로 놓는 것이 이미 놓인 것과 겹치지 않는 가장 가까운 오른쪽 자리. 건물처럼 큰 대역도 비켜 선다.
export function freeSpot(placed, dimensions, gap = 0.3) {
  const [width, depth] = (dimensions || [1, 1]).map(Number);
  const hits = (x) => placed.some((item) => {
    const { low, high } = footprint(item);
    return x - width / 2 < high[0] + gap && x + width / 2 > low[0] - gap && depth / 2 > low[1] - gap && -depth / 2 < high[1] + gap;
  });
  let x = 0;
  while (hits(x)) x += 0.5;
  return [x, 0];
}

export function sceneSubjects(placed) {
  const boxes = placed.map((item, index) => ({ id: assetId(index), ...footprint(item) }));
  const subjects = Object.fromEntries(boxes.map((box) => [box.id, subjectOf(box.low, box.high)]));
  if (boxes.length) {
    const low = [0, 1, 2].map((axis) => Math.min(...boxes.map((box) => box.low[axis])));
    const high = [0, 1, 2].map((axis) => Math.max(...boxes.map((box) => box.high[axis])));
    subjects.scene = subjectOf(low, high);
  }
  return subjects;
}

// 피사체가 화면을 꽉 채우는 거리. 프레이밍 거리 1.0이 이 거리다.
export function fitDistance(lens, subject, aspect = ASPECT) {
  return Math.max((subject.width * lens) / SENSOR_MM, (subject.height * lens * aspect) / SENSOR_MM);
}

export function focusOf(cut, subjects) {
  return subjects[cut.focus || "scene"] || subjects.scene || null;
}

export function isMoving(cut) {
  const framing = cut.framing || {};
  return Object.keys(framing).some((key) => key.endsWith("End")) || (cut.lensEnd != null && cut.lensEnd !== cut.lens);
}

export function cameraAt(cut, subjects, fraction = 0, aspect = ASPECT) {
  const subject = focusOf(cut, subjects);
  if (!subject) return null;
  const framing = cut.framing || {};
  const ease = cut.ease || "inout";
  const lens = smooth(...pair(cut, "lens", 35), fraction, ease);
  const distance = smooth(...pair(framing, "distance", DEFAULTS.distance), fraction, ease) * fitDistance(lens, subject, aspect);
  const azimuth = (smooth(...pair(framing, "azimuth", DEFAULTS.azimuth), fraction, ease) * Math.PI) / 180;
  const [cx, cy] = subject.center;
  const offset = framing.targetOffset || [0, 0];
  return {
    lens,
    position: [
      cx + distance * Math.sin(azimuth),
      cy - distance * Math.cos(azimuth),
      subject.baseZ + smooth(...pair(framing, "height", DEFAULTS.height), fraction, ease) * subject.height,
    ],
    target: [
      cx + offset[0] * subject.radius,
      cy + offset[1] * subject.radius,
      subject.baseZ + smooth(...pair(framing, "targetHeight", DEFAULTS.targetHeight), fraction, ease) * subject.height,
    ],
  };
}

// 지도에서 끈 점을 방향·거리 값으로 되돌린다. 방향은 원래 값과 가장 가까운 각을 골라
// 궤도 촬영이 반대쪽으로 한 바퀴 돌지 않게 한다.
export function aimFromPoint(cut, subjects, point, handle = "start", aspect = ASPECT) {
  const subject = focusOf(cut, subjects);
  if (!subject) return cut;
  const framing = { ...(cut.framing || {}) };
  const end = handle === "end";
  const lens = end ? pair(cut, "lens", 35)[1] : pair(cut, "lens", 35)[0];
  const dx = point[0] - subject.center[0];
  const dy = point[1] - subject.center[1];
  const previous = end ? pair(framing, "azimuth", DEFAULTS.azimuth)[1] : pair(framing, "azimuth", DEFAULTS.azimuth)[0];
  const raw = (Math.atan2(dx, -dy) * 180) / Math.PI;
  const turn = ((((raw - previous) % 360) + 540) % 360) - 180;
  const azimuth = Math.round(previous + turn);
  const distance = Math.round(clamp(Math.hypot(dx, dy) / fitDistance(lens, subject, aspect), LIMITS.distance) * 20) / 20;
  if (end) {
    framing.azimuthEnd = azimuth;
    framing.distanceEnd = distance;
  } else {
    framing.azimuth = azimuth;
    framing.distance = distance;
  }
  return { ...cut, framing };
}

export function draftFromPreset(preset) {
  return clone(preset?.shots || []).map((shot) => ({ ...shot, framing: shot.framing || {} }));
}

export function moveItem(list, from, to) {
  const next = [...list];
  const [item] = next.splice(from, 1);
  next.splice(Math.max(0, Math.min(next.length, to)), 0, item);
  return next;
}

export function nextShotId(cuts) {
  const used = new Set(cuts.map((cut) => cut.id));
  for (let index = 1; ; index += 1) {
    const id = `s${String(index).padStart(2, "0")}`;
    if (!used.has(id)) return id;
  }
}

export function totalSeconds(cuts) {
  return cuts.reduce((sum, cut) => sum + Number(cut.seconds || 0), 0);
}

export function setSeconds(cut, seconds) {
  return { ...cut, seconds: Math.round(clamp(Number(seconds) || 0, LIMITS.seconds) * 10) / 10 };
}

// 사람이 고친 장면과 컷을 엔진 요청 하나로 바꾼다. 프리셋 그대로면 컷을 보내지 않아
// 기록에 "편집"이 붙지 않는다.
export function buildPrevizRequest({ placed = [], cuts = [], preset = null, settings = {}, description = "" }) {
  if (!placed.length) throw new Error("장면에 3D 에셋이나 대역을 하나 이상 놓아 주세요.");
  if (!cuts.length) throw new Error("컷이 하나 이상 있어야 합니다.");
  const assets = placed.map((item, index) => {
    const number = (value, name, [low, high]) => {
      const parsed = Number(String(value ?? "").trim() || 0);
      if (!Number.isFinite(parsed) || parsed < low || parsed > high) {
        throw new Error(`${index + 1}번째 ${item.standin ? "대역" : "에셋"}의 ${name} 값을 확인해 주세요.`);
      }
      return parsed;
    };
    const placement = {
      id: assetId(index),
      position: [number(item.x, "x", [-1000, 1000]), number(item.y, "y", [-1000, 1000]), 0],
      yaw: number(item.yaw, "회전", [-360, 360]),
    };
    if (item.standin) {
      // 대역은 크기 배율 대신 미터 치수를 보낸다. 엔진이 부품을 그 치수로 늘인다.
      const size = SIZE_LABELS.map((name, axis) => number(item.dimensions?.[axis], name, LIMITS.size));
      return { ...placement, standin: item.standin, size };
    }
    return {
      ...placement,
      source: { jobId: item.jobId, assetId: item.assetId },
      scale: number(item.scale ?? 1, "크기", [0.01, 100]),
    };
  });
  const edited = JSON.stringify(cuts) !== JSON.stringify(draftFromPreset(preset));
  const [width, height] = String(settings.resolution || "960x540").split("x").map(Number);
  return {
    recipe: "previz",
    params: {
      ...(description.trim() ? { description: description.trim() } : {}),
      preset: preset?.id, assets, ...(edited ? { shots: clone(cuts) } : {}),
      renderer: settings.renderer || "eevee", width, height, fps: Number(settings.fps || 12),
      aux: settings.aux || "keys", clay: settings.clay !== false, animatic: settings.animatic !== false,
    },
  };
}
